source/opt/ir_context.cpp

```diff
@@ -218,6 +220,8 @@ Instruction* IRContext::KillInst(Instruction* inst) {
 void IRContext::CollectNonSemanticTree(
     Instruction* inst, std::unordered_set<Instruction*>* to_kill) {
   if (!inst->HasResultId()) return;
+  // Debug[No]Line result id is not used, so we are done
+  if (inst->IsDebugLineInst()) return;
```

Comment:
I think this is wrong.
If it is a non-semantic line instruction, we should put it in `to_kill`.
In addition, [DebugLine](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/nonsemantic/NonSemantic.Shader.DebugInfo.100.html#DebugLine) has an operand `Source` that must be a `DebugSource`. We should put the `DebugSource` in `to_kill` as well.

So .. I think we have to remove this line.
Correct me if I am wrong.

Reply:
I think you are thinking about this in the reverse way.

Here is the comment for CollectNonSemanticTree:

```
  // Collects the non-semantic instruction tree that uses |inst|'s result id
  // to be killed later.
```

We know there are no users of Debug[No]Line result ids. If inst is a Debug[No]Line, the rest of the algorithm won't do anything, so we may as well return here. Right? This is just an efficiency, a short cut. If I remove this, it won't change any results, the optimizer will just run slightly slower.

Yes, Debug[No]Line uses DebugSource. If inst is a DebugSource, then all its users, including any and all Debug[No]Line, will be added to to_kill by the following loop.

Reply:
Oops sorry. I was confused. You are right.